app/handler/transactional_complaint.py

```diff
@@ -0,0 +1,229 @@
+import uuid
+from abc import ABC, abstractmethod
+from io import BytesIO
+from mailbox import Message
+from typing import Optional
+
+from app import s3
+from app.config import (
+    ALERT_COMPLAINT_REPLY_PHASE,
+    ALERT_COMPLAINT_TO_USER,
+    ALERT_COMPLAINT_FORWARD_PHASE,
+)
+from app.email import headers
+from app.email_utils import (
+    get_header_unicode,
+    parse_full_address,
+    save_email_for_debugging,
+    to_bytes,
+    render,
+    send_email_with_rate_control,
+)
+from app.log import LOG
+from app.models import (
+    User,
+    Alias,
+    DeletedAlias,
+    DomainDeletedAlias,
+    Contact,
+    TransactionalComplaint,
+    Phase,
+    TransactionalComplaintState,
+    RefusedEmail,
+)
+
+
+class TransactionalComplaintOrigin(ABC):
+    @classmethod
+    @abstractmethod
+    def get_original_message(cls, message: Message) -> Optional[Message]:
+        pass
+
+    @classmethod
+    @abstractmethod
+    def name(cls):
+        pass
+
+
+class TransactionalYahooOrigin(TransactionalComplaintOrigin):
+    @classmethod
+    def get_original_message(cls, message: Message) -> Optional[Message]:
+        wanted_part = 6
+        for part in message.walk():
+            wanted_part -= 1
+            if wanted_part == 0:
+                return part
+        return None
+
+    @classmethod
+    def name(cls):
+        return "yahoo"
+
+
+class TransactionalHotmailOrigin(TransactionalComplaintOrigin):
+    @classmethod
+    def get_original_message(cls, message: Message) -> Optional[Message]:
+        wanted_part = 3
+        for part in message.walk():
+            wanted_part -= 1
+            if wanted_part == 0:
+                return part
+        return None
+
+    @classmethod
+    def name(cls):
+        return "hotmail"
+
+
+def handle_hotmail_complaint(message: Message) -> bool:
+    return handle_complaint(message, TransactionalHotmailOrigin())
+
+
+def handle_yahoo_complaint(message: Message) -> bool:
+    return handle_complaint(message, TransactionalYahooOrigin())
+
+
+def find_alias_with_address(address: str) -> Optional[Alias]:
+    return (
+        Alias.get_by(email=address)
+        or DeletedAlias.get_by(email=address)
+        or DomainDeletedAlias.get_by(email=address)
+    )
+
+
+def handle_complaint(message: Message, origin: TransactionalComplaintOrigin) -> bool:
+    original_message = origin.get_original_message(message)
+
+    try:
+        _, to_address = parse_full_address(
+            get_header_unicode(original_message[headers.TO])
+        )
+        _, from_address = parse_full_address(
+            get_header_unicode(original_message[headers.FROM])
+        )
+    except ValueError:
+        saved_file = save_email_for_debugging(message, "FromParseFailed")
+        LOG.w("Cannot parse from header. Saved to {}".format(saved_file or "nowhere"))
+        return True
```

Comment:
when there's an error, maybe we can return False so the email will be forwarded to postmaster and we can then take a look at the email content?

Reply:
:+1: